### PR TITLE
Feature Gates without a Config File

### PR DESCRIPTION
**Pull request by tzaffi**
Addressing the pyteal.ini portion of https://github.com/algorand/pyteal/issues/677 by **removing it** as well as proposing a way to address https://github.com/algorand/pyteal/issues/601

Please see `examples/applications/sourcemap.py` for usage example.

## Regarding Feature Gates Issue [601](https://github.com/algorand/pyteal/issues/601)

`feature_gates.FeatureGates` provides a mechanism for defining global booleans that are loaded before the rest of PyTeal is. An experimental code path could directly use these globals via `FeatureGates.theFeature_isOn()` or I can also imagine a decorator such as `@FeatureGate.decorate_theFeature_isOn` which actually runs an `assert FeatureGates.theFeature_isOn()` as part of the wrapped method. To avail the feature in a particular build, one could call `FeatureGate.set_theFeature_isOn(True)` and to officially release (without removing the decorators) is as simple as changing the default defined in `FeatureGates._gates.theFeature_isOn`

>Optionally, providing an explicit opt-in mechanism to use experimental features.

That's available via -for example- `FeatureGate.set_theFeature_isOn(True)`

### CODE REVIEW

File: examples/application/sourcemap.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+"""
+This example shows how to use the sourcemap feature to generate a source map for a PyTeal program.
```
Comment (tzaffi) [inline examples/application/sourcemap.py:2]:
```suggestion
This example shows how to use the source map feature to for a PyTeal program.
```

File: feature_gates/__init__.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    def set(cls, feature: str, gate: bool) -> None:
+        if feature not in cls._features:
+            raise ValueError(
+                f"Cannot set unknown {feature=} (available features: {cls._features}))"
+            )
+        setattr(cls._gates, feature, gate)
+
+
+for feature in FeatureGates._features:
+    FeatureGates._make_feature(feature)
```
Comment (tzaffi) [inline feature_gates/__init__.py:66]: We implicitly create the following methods:

```
@classmethod
def FeatureGates.sourcemap_enabled(cls):

@classmethod
def FeatureGates.sourcemap_debug(cls):

@classmethod
def FeatureGates.set_sourcemap_enabled(cls):

@classmethod
def FeatureGates.set_sourcemap_debug(cls):
```

File: feature_gates/__init__.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    We follow the following convention:
+
+    feature_property: True/False
+    """
+
+    sourcemap_enabled: bool
+    sourcemap_debug: bool
+
+
+class FeatureGates:
```
Comment (tzaffi) [inline feature_gates/__init__.py:19]: I can split this out of the `__init__.py` if I get a comment in that direction.

File: pyteal/__init__.py [hunk 4]
```diff
@@ -1,69 +1,69 @@
 from pyteal.ast import *
 from pyteal.ast import __all__ as ast_all
-from pyteal.pragma import pragma
```
Comment (tzaffi) [inline pyteal/__init__.py:3]: Originally I needed to touch this file in order to "load()" the `FeatureGates` from the `.ini` if there was one. In touching the file, I re-sorted the imports.  I can revert to the original if I get a request.

File: examples/application/sourcemap.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+To enable the source mapping (at the time of this writing 3/13/2023)
+one must import `FeatureGate` from `feature_gates` and call `FeatureGate.set_sourcemap_enabled(True)`.
+This import must occur before any object from PyTeal is imported, as PyTeal will
+default to its own sourcemap configuration if not set beforehand.
+"""
+
+from feature_gates import FeatureGates
+
+
```
Comment (bbroder-algo) [inline examples/application/sourcemap.py:12]: wouldn't it be less confusing if pyteal imports and feature setting were at the top
like 
```
from feature_gates import FeatureGates
FeatureGates.verbose = True
FeatureGates.set_sourcemap_enabled(True)
import pyteal
```

File: feature_gates/__init__.py (hunk 2 above)
Comment (bbroder-algo) [reply feature_gates/__init__.py:66]: cool

File: tests/unit/feature_gates_test.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+@pytest.mark.serial
+def test_feature_gates():
+    with pytest.raises(ValueError, match="Unknown feature='foo'"):
+        FeatureGates.get("foo")
+
+    with pytest.raises(ValueError, match="Cannot set unknown feature='foo'"):
+        FeatureGates.set("foo", 42)
+
+    # presently, source mapping features are off by default:
```
Comment (bbroder-algo) [inline tests/unit/feature_gates_test.py:14]: This feels like a test of the default and not a test of the feature gates. 
maybe just pull False and True out into a local variable `default_sourcemap_gate : bool = FeatureGates.get("sourcemap_enabled")` and then the asserts become `assert FeatureGates.get("sourcemap_enabled") is default_sourcemap_gate` or `not default_sourcemap_gate`

File: tests/unit/sourcemap_monkey_raises_test.py [hunk 7]
```diff
@@ -1,31 +1,30 @@
...
-    with pytest.raises(SourceMapDisabledError) as smde:
-        router.compile(
-            version=6,
-            optimize=OptimizeOptions(scratch_slots=True),
-            with_sourcemaps=True,
-        )
-
-    assert "pyteal.ini" in str(smde.value)
-    assert "1337" not in str(smde.value)
+    with sourcemap_disabled():
```
Comment (bbroder-algo) [inline tests/unit/sourcemap_monkey_raises_test.py:22]: nice

**Review by ahangsu — COMMENTED** [review]
finished my first pass, looks generally good to me

File: feature_gates/__init__.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    Automagically, a new setter and getter will be available for the feature gate.
+    For example, given feature `foo_on` by adding _FeatureGatesConfig.foo_on,
+    the following methods will be available:
+    * FeatureGates.foo_on() -> bool
+    * FeatureGates.set_foo_on(gate: bool)
+    """
+
+    # default values for feature gates:
+    _gates: Final[_FeatureGatesConfig] = _FeatureGatesConfig(
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply feature_gates/__init__.py:33]: This is intentionally `Final` (but confusing). I'm okay with swapping out the fields of the `_FetureGatesConfig` **dataclass** (it isn't frozen), but not the object itself.

[the remark this answers is not in the document]
Comment (ahangsu) [reply feature_gates/__init__.py:33]: oh I get it... you are not allowing to set `_gates`, but allowing setting members of the `_gates`... resolving.

File: examples/application/sourcemap.py (hunk 5 above)
Comment (tzaffi) [reply examples/application/sourcemap.py]: 63ada3df475de1213dbc55aca6a8713ffc03a98c

File: tests/unit/feature_gates_test.py (hunk 6 above)
Comment (tzaffi) [reply tests/unit/feature_gates_test.py]: bcf1d4704e2cbd8675bb72291e4f635472b32aec

File: pyteal/stack_frame.py [hunk 9]
```diff
@@ -226,25 +212,15 @@ class NatalStackFrame:
...
-        * Frames are useful in an "all or nothing" capacity. For example, in preparing
-            for a source mapping, it would be error prone to generate frames for
-            a subset of analyzed PyTeal
-        * Setting `_force_refresh = True` will cause a read from the file system every
-            time Frames are initialized and will result in significant performance degredation
-        """
-        if _force_refresh:
-            cls._no_stackframes = _sourcmapping_is_off()
+    def sourcemapping_is_off(cls) -> bool:
+        return not FeatureGates.sourcemap_enabled()  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply pyteal/stack_frame.py]: d2a1cbdc4318a084b8d71088d65dadaae22fa08f

File: pyteal/stack_frame.py [hunk 10]
```diff
@@ -226,25 +212,15 @@ class NatalStackFrame:
...
-        """
-        if _force_refresh:
-            cls._no_stackframes = _sourcmapping_is_off()
+    @staticmethod
+    def sourcemapping_is_off() -> bool:
+        return not FeatureGates.sourcemap_enabled()  # type: ignore[attr-defined]
 
-        return cls._no_stackframes
+    @classmethod
+    def _debugging(cls) -> bool:
```
Comment (ahangsu) [inline pyteal/stack_frame.py:222]: nit: same here
```suggestion
    @staticmethod
    def _debugging() -> bool:
```

**Review by ahangsu — APPROVED** [review]
lgtm

**Review by bbroder-algo — APPROVED** [review]

**Pull request merged**